File: tests/architecture-audit-checker/go/loader/loader.go
// Package loader reads the six audit deliverable Markdown files from the
// architecture-audit spec directory into a models.Deliverables value.
//
// Missing files are tolerated: a file that does not exist surfaces as an
// empty string in the corresponding field. Any other read error
// (permission denied, I/O error) is returned to the caller.
package loader

import (
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"

	"github.com/robinhood/architecture-audit-checker-go/models"
)

// deliverableFiles enumerates the six expected Markdown deliverables in the
// order they appear in the architecture-audit design document. The closed
// set is enforced by Property 1; this loader never reads any other file.
var deliverableFiles = []string{
	"audit-report.md",
	"dependency-map.md",
	"migration-plan.md",
	"compatibility-strategy.md",
	"preservation-inventory.md",
	"rebuild-recommendations.md",
}

// Load reads the six audit deliverables from specDir and returns a
// populated models.Deliverables. specDir is typically
// .kiro/specs/architecture-audit/ relative to the repository root.
//
// A missing file (os.ErrNotExist) is tolerated and surfaces as an empty
// string in the corresponding field. Any other error is returned wrapped
// with the offending path.
func Load(specDir string) (models.Deliverables, error) {
	contents := make(map[string]string, len(deliverableFiles))
	for _, name := range deliverableFiles {
		full := filepath.Join(specDir, name)
		data, err := os.ReadFile(full)
		switch {
		case err == nil:
			contents[name] = string(data)
		case errors.Is(err, fs.ErrNotExist):
			contents[name] = ""
		default:
			return models.Deliverables{}, fmt.Errorf("loader: read %s: %w", full, err)
		}
	}
	return models.Deliverables{
		AuditReport:            contents["audit-report.md"],
		DependencyMap:          contents["dependency-map.md"],
		MigrationPlan:          contents["migration-plan.md"],
		CompatibilityStrategy:  contents["compatibility-strategy.md"],
		PreservationInventory:  contents["preservation-inventory.md"],
		RebuildRecommendations: contents["rebuild-recommendations.md"],
	}, nil
}

File: tests/architecture-audit-checker/csharp/src/AuditChecker/Loader.cs
// Loader for the six Markdown deliverables under
// .kiro/specs/architecture-audit/. The loader is read-only, takes the
// directory path as input, tolerates missing files (returns empty strings),
// and never writes outside its own track.

using Robinhood.Audit.Models;

namespace Robinhood.Audit;

/// <summary>
/// Reads the six Markdown deliverables from a spec directory into a
/// <see cref="Deliverables"/> record. Missing files yield empty strings so
/// the 17 Correctness Properties run end-to-end and report structural gaps
/// rather than crashing at I/O time.
/// </summary>
public static class Loader
{
    /// <summary>
    /// Filenames of the six audit deliverables, in the order they appear in
    /// the <see cref="Deliverables"/> record's constructor.
    /// </summary>
    public static readonly IReadOnlyList<string> DeliverableFilenames = new[]
    {
        "audit-report.md",
        "dependency-map.md",
        "migration-plan.md",
        "compatibility-strategy.md",
        "preservation-inventory.md",
        "rebuild-recommendations.md",
    };

    /// <summary>
    /// Loads the six deliverable Markdown files from <paramref name="specDir"/>.
    /// </summary>
    /// <param name="specDir">
    /// Absolute or workspace-relative path of the
    /// <c>.kiro/specs/architecture-audit/</c> directory.
    /// </param>
    /// <returns>
    /// A <see cref="Deliverables"/> populated with each file's text. Any file
    /// that does not exist on disk is represented by an empty string.
    /// </returns>
    public static Deliverables Load(string specDir)
    {
        ArgumentNullException.ThrowIfNull(specDir);

        return new Deliverables(
            AuditReport: ReadOrEmpty(specDir, "audit-report.md"),
            DependencyMap: ReadOrEmpty(specDir, "dependency-map.md"),
            MigrationPlan: ReadOrEmpty(specDir, "migration-plan.md"),
            CompatibilityStrategy: ReadOrEmpty(specDir, "compatibility-strategy.md"),
            PreservationInventory: ReadOrEmpty(specDir, "preservation-inventory.md"),
            RebuildRecommendations: ReadOrEmpty(specDir, "rebuild-recommendations.md"));
    }

    private static string ReadOrEmpty(string specDir, string filename)
    {
        var path = Path.Combine(specDir, filename);
        return File.Exists(path) ? File.ReadAllText(path) : string.Empty;
    }
}
